src/lustre/lustreAstNormalizer.ml

```diff
@@ -48,194 +50,383 @@
+
      @author Andrew Marmaduke *)
 
 open Lib
 
 module A = LustreAst
 module AH = LustreAstHelpers
+module Ctx = TypeCheckerContext
+module Chk = LustreTypeChecker
 
-(** Map for types with identifiers as keys *)
-module IMap = struct
+module StringMap = struct
   include Map.Make(struct
-    type t = LustreAst.ident
+    type t = string
     let compare i1 i2 = String.compare i1 i2
   end)
   let keys: 'a t -> key list = fun m -> List.map fst (bindings m)
 end
 
 type generated_identifiers = {
-    locals : LustreAst.expr IMap.t;
-    oracles : LustreAst.ident list
+  node_args : (string (* abstracted variable name *)
+    * bool (* whether the variable is constant *)
+    * LustreAst.lustre_type
+    * LustreAst.expr)
+    list;
+  locals : (bool (* whether the variable is ghost *)
+    * LustreAst.lustre_type
+    * LustreAst.expr)
+    StringMap.t;
+  oracles : (string * LustreAst.expr) list;
+  propagated_oracles : (string * string) list;
+  calls : (Lib.position (* node call position *)
+    * (string list) (* oracle inputs *)
+    * (string list) (* abstracted inputs *)
+    * LustreAst.expr (* condition expression *)
+    * LustreAst.expr (* restart expression *)
+    * string (* node name *)
+    * (LustreAst.expr list) (* node arguments *)
+    * (LustreAst.expr list option)) (* node argument defaults *)
+    list
 }
 
 let pp_print_generated_identifiers ppf gids =
-  let locals_list = IMap.bindings gids.locals |> List.rev in
-  let pp_print_local = pp_print_pair
+  let locals_list = StringMap.bindings gids.locals 
+    |> List.map (fun (x, (y, z, w)) -> x, y, z, w)
+  in let pp_print_local ppf (id, b, ty, e) = Format.fprintf ppf "(%a, %b, %a, %a)"
+    Format.pp_print_string id
+    b
+    LustreAst.pp_print_lustre_type ty
+    LustreAst.pp_print_expr e
+  in let pp_print_oracle = pp_print_pair
     Format.pp_print_string
-    A.pp_print_expr
-    " = "
-  in
-  let pp_print_oracle = Format.pp_print_string in
-  Format.fprintf ppf "%a\n%a"
+    LustreAst.pp_print_expr
+    ":"
+  in let pp_print_call = (fun ppf (pos, vars, oracles, cond, restart, ident, args, defaults) ->
+    Format.fprintf ppf 
+      "%a: %a = call(%a,(restart %a every %a)(%a;%a),%a)"
+      pp_print_position pos
+      (pp_print_list Format.pp_print_string ",@") vars
+      A.pp_print_expr cond
+      Format.pp_print_string ident
+      A.pp_print_expr restart
+      (pp_print_list A.pp_print_expr ",@ ") args
+      (pp_print_list Format.pp_print_string ",@ ") oracles
+      (pp_print_option (pp_print_list A.pp_print_expr ",@")) defaults)
+  in Format.fprintf ppf "%a\n%a\n%a\n%a"
     (pp_print_list pp_print_oracle "\n") gids.oracles
     (pp_print_list pp_print_local "\n") locals_list
+    (pp_print_list pp_print_local "\n") gids.node_args
+    (pp_print_list pp_print_call "\n") gids.calls
 
 
 (* [i] is module state used to guarantee newly created identifiers are unique *)
 let i = ref 0
 
+let is_input_const = ref StringMap.empty
```

Comment:
You should reset `is_input_const` at the beginning of `normalize` so that the function can be called more times if needed.

Reply:
Well, since `normalize` is a recursive function right now, you may need to split it into two first.

Reply:
I don't think it is needed because `is_input_const` is only tracking if the the inputs to a node are constant (information I hoped could be extracted from the type checker context, but it is not being tracked). This information should be static across calls to `normalize`

Reply:
Although it would be better to write an auxiliary function that populates it once and only once

Reply:
> This information should be static across calls to normalize

Provided that the input program is the same.

Although the current pipeline reads only one input program and then process it, I would like to have the flexibility of compiling different lustre programs in the same run of the tool.

Reply:
Gotcha, will do